### ISSUE TITLE
Question - Docker Compose Setup

### ISSUE THREAD

**Issue by samcro1967** [fmartinou/whats-up-docker#158]
I am not seeing any updates being made when I do a first run.  Looks like it is having issues parsing docker-compse.yml???

Docker Compose
```
  whatsupdocker:
    <<: *logging
    hostname: whatsupdocker
    image: fmartinou/whats-up-docker
    container_name: whatsupdocker
    restart: always
    environment:
      - TZ=${TZ}
      - WUD_SERVER_PORT=8841
      - WUD_REGISTRY_CUSTOM_URL=http://local.registry:5000
      - WUD_REGISTRY_GHCR_USERNAME=<redacted>
      - WUD_REGISTRY_GHCR_TOKEN=ghp_<redacted>
      - WUD_REGISTRY_GCR_CLIENTEMAIL=<redacted>
      - WUD_REGISTRY_GCR_PRIVATEKEY=$GCR_PRIVATEKEY
      - WUD_REGISTRY_HUB_LOGIN=<redacted>
      - WUD_REGISTRY_HUB_TOKEN=<redacted>
      - WUD_WATCHER_LOCAL_CRON=0 8 * * *
      - WUD_LOG_LEVEL=debug #error info debug trace
      - WUD_TRIGGER_DOCKERCOMPOSE_WUD_FILE=/wud/docker-compose.yml
      - WUD_TRIGGER_DOCKERCOMPOSE_WUD_BACKUP=true
      - WUD_TRIGGER_DOCKERCOMPOSE_WUD_PRUNE=true
      - WUD_TRIGGER_DOCKERCOMPOSE_WUD_ONCE=false
      - WUD_TRIGGER_DOCKERCOMPOSE_WUD_MODE=batch
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - "${docker}/whatsupdocker:/store"
      - /etc/localtime:/etc/localtime:ro
      - "${docker}/docker-compose.yml:/wud/docker-compose.yml"
    ports:
      - 8841:8841
```

**Logs**
```
08:45:32.451 DEBUG whats-up-docker/trigger.dockercompose.wud: Run trigger batch
08:45:32.452  INFO whats-up-docker/watcher.docker.local: Cron finished (117 containers watched, 0 errors, 25 available updates)
08:45:32.558  WARN whats-up-docker/trigger.dockercompose.wud: Error (Cannot read properties of undefined (reading 'error'))
08:45:32.558 DEBUG whats-up-docker/trigger.dockercompose.wud: Cannot read properties of undefined (reading 'error')
    TypeError: Cannot read properties of undefined (reading 'error')
        at getComposeFile (/home/node/app/triggers/providers/dockercompose/Dockercompose.js:15:18)
        at async Dockercompose.triggerBatch (/home/node/app/triggers/providers/dockercompose/Dockercompose.js:55:25)
        at async Dockercompose.handleContainerReports (/home/node/app/triggers/providers/Trigger.js:75:17)
```

**Comment by fmartinou** [comment]
Either wud hasn't found the file (but your config looks ok)

Or wud has issues with parsing the yaml
=> maybe is the cause the use of yaml anchors/aliases (`<<: *logging`)?
Can you retry without them?

**Comment by samcro1967** [comment]
I commented out that line for all containers and the anchor definition itself.  Also confirmed with docker exec that /wud/docker-compose.yml is mounted and has the correct contents.

**Comment by samcro1967** [comment]
Is there anyway I can tune logging to show which line in docker compose it is not parsing properly?

**Comment by fmartinou** [comment]
> Is there anyway I can tune logging to show which line in docker compose it is not parsing properly?

Unfortunately not; I need to change the code (I will because it's important for troubleshooting).
For the moment, can you share you full compose file so that I can test it on my end?

**Comment by samcro1967** [comment]
See if you can pull it from here.  Let me know once you have it so I can remove it.

[Link](https://hastebin.samcro1967.net/pukoqodupi.bash)

**Comment by fmartinou** [comment]
I get an `Access denied`

**Comment by samcro1967** [comment]
Hmmm...ok  Pastebin will not let me upload it.  Says potentially offensive content.  LOL.  See if this works.  I will delete the post after you grab it.
[docker-compose.7z.txt](https://github.com/fmartinou/whats-up-docker/files/9496683/docker-compose.7z.txt)

**Comment by fmartinou** [comment]
I get it 👍 

You can delete it.
Thanks!

**Comment by fmartinou** [comment]
```
Error when parsing the docker-compose yaml file (Map keys must be unique at line 447, column 5:
    
          test: [NONE]
        devices:
        ^
    )
```

Indeed, you have the `devices`prop twice

=> I'll change the code to improve the logs when this kind of errors occurs.

**Comment by samcro1967** [comment]
That has been there for years.  Apparently docker compose has some validity verification issues.  Have updated and restarted.  Will monitor and see what happens on the next run in a few hours.

**Comment by samcro1967** [comment]
I think I have encountered another issue in parsing docker-compose as I am not seeing any update attempts.  Here is the end of the log after a run.   Is there a particular tool you use to validate a compose file to find potential errors that are missed by docker compose itself until logging can be enhanced?

```
15:02:30.367  INFO whats-up-docker/watcher.docker.local: Cron finished (36 containers watched, 0 errors, 3 available updates)
15:02:30.566  WARN whats-up-docker/trigger.dockercompose.wud: Error (Cannot read properties of undefined (reading 'error'))
```

**Comment by fmartinou** [comment]
I'm not validating the content of the file as a valid "docker-Compose"; I'm just parsing it as a "regular" yaml file.
If the file content is not "yaml" syntactically valid; it throws an error.

I guess that the docker-compose parser (python based) is more permissive.

For example, on the parser I'm using, those options are enabled by default:
- `strict` : When parsing, do not ignore errors required by the YAML 1.2 spec, but caused by unambiguous content.
- `uniqueKeys`: Whether key uniqueness is checked

**Comment by samcro1967** [comment]
Hmm...ok.  I have used online yaml validators and they give my docker compose a clean bill of health from a yaml perspective.  Will wait until logging is enhanced to shed some more light on where the issue lies.  Thanks for your help.  Already getting value out of whatsupdocker even without the docker compose automation.  Like what I am seeing thus far!

**Comment by fmartinou** [comment]
Can you give a try to the version `fmartinou/whats-up-docker:fix_improve_docker_compose_yaml_error_logging` ?
It should output better logs when an docker-compose parsing error is occurring.

**Comment by samcro1967** [comment]
Yeah, that worked and helped.  Was able to determine the parsing fails when you comment out a service block that is greater than 1024 characters in length across multiple lines.  Removed the 2 causing issues and is now parsing properly.

**Comment by samcro1967** [comment]
Looks like when you parse docker compose you also remove all comments?

**Comment by fmartinou** [comment]
Yeah, that's true 😞 

For now, the compose file is making a round trip (YAML -> Javascript -> YAML) so all comments and empty lines are lost.

=> I reopen your issue to take the opportunity to improve that part.

**Comment by fmartinou** [comment]
Can you pull the latest `fmartinou/whats-up-docker:fix_improve_docker_compose_yaml_error_logging` and test?

The compose file formatting (comments & empty lines) should be preserved.

**Comment by samcro1967** [comment]
Can confirm comments are now being preserved.

**Comment by fmartinou** [comment]
Available in `4.20.3`
